nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -1242,14 +1305,16 @@ impl DataStore {
                             "found_gen" => ?found.updater_gen,
                             "locked_gen" => ?locked_gen,
                         );
-                        debug_assert_ne!(actual_id, updater_id);
                         Err(Error::internal_error(
                             "attempted to release a lock held by another saga! this is a bug!",
                         ))
                     },
+                    Some(_) =>  Err(Error::conflict(
+                        "attempted to commit an instance update, but the state generation has advanced!"
+                    )),
```

Comment:
related to the above: this case is mildly at odds with the comment on line 1295, since it's not a programmer error--another saga could have bumped the generation number by kicking off a migration saga that ran concurrently with this update. I think this is just a comment bug, though; there's nothing better to do here than to return this error, let the update unwind, and then try again later.

(Incidentally, if this happens, will the unwinding update saga reschedule itself to try to update again immediately? I'm not sure I see anything like that happening in `siu_unbecome_updater`. Do we just rely on the periodic timer for this case?)

Reply:
Ah, yeah, I think unwinding once the lock has been acquired and we've determined that we *do* actually need to perform an update should probably kick the background task, good call.

Reply:
Hmm, actually, I don't think that's true. This error is returned in the match
arm that checks that the updater ID is not the expected one. The next match arm
handles the case where the instance's state generation has changed, and returns
`Error::Conflict` instead.

We should probably change the error type here to something that the caller can
easily match on to determine if a new saga should be scheduled...though honestly
any time an instance update saga unwinds, if we needed to perform updates, a new
update saga should be scheduled, regardless of why we unwound...